apollo-query-graph/src/extract_subgraphs_from_supergraph.rs

```diff
@@ -0,0 +1,1703 @@
+        add_federation_operations(subgraph, federation_spec_definition);
+        if validate_extracted_subgraphs {
+            let Some(diagnostics) = subgraph.schema.schema.validate().err() else {
+                continue;
+            };
+            // TODO: Implement maybeDumpSubgraphSchema() for better error messaging
+            if is_fed_1 {
+                // See message above about Fed 1 supergraphs
+                todo!()
+            } else {
+                return Err(
+                    ErrorCode::InvalidFederationSupergraph.definition().err(
+                        format!(
+                            "Unexpected error extracting {} from the supergraph: this is either a bug, or the supergraph has been corrupted.\n\nDetails:\n{}",
+                            subgraph.name,
+                            diagnostics.to_string_no_color()
+                        ),
+                        None,
+                    ).into()
+                );
+            }
+        }
+    }
+
+    Ok(subgraphs)
+}
+
+fn validate_supergraph(
+    supergraph_schema: &Schema,
+) -> Result<
+    (
+        FederationSchemaRef<OptionLinksMetadata>,
+        &'static LinkSpecDefinition,
+        &'static JoinSpecDefinition,
+    ),
+    FederationError,
+> {
+    let supergraph_schema = FederationSchemaRef::<OptionLinksMetadata>::new(supergraph_schema)?;
+    let Some(metadata) = supergraph_schema.metadata() else {
+        return Err(ErrorCode::InvalidFederationSupergraph
+            .definition()
+            .err("Invalid supergraph: must be a core schema".to_owned(), None)
+            .into());
+    };
+    let link_spec_definition = metadata.link_spec_definition();
+    let Some(join_link) = metadata.for_identity(&Identity::join_identity()) else {
+        return Err(ErrorCode::InvalidFederationSupergraph
+            .definition()
+            .err(
+                "Invalid supergraph: must use the join spec".to_owned(),
+                None,
+            )
+            .into());
+    };
+    let Some(join_spec_definition) = JOIN_VERSIONS.find(&join_link.url.version) else {
+        return Err(ErrorCode::InvalidFederationSupergraph.definition().err(
+            format!(
+                "Invalid supergraph: uses unsupported join spec version {} (supported versions: {})",
+                JOIN_VERSIONS.versions().map(|v| v.to_string()).collect::<Vec<String>>().join(", "),
+                join_link.url.version,
+            ),
+            None,
+        ).into());
+    };
+    Ok((
+        supergraph_schema,
+        link_spec_definition,
+        join_spec_definition,
+    ))
+}
+
+fn collect_empty_subgraphs<'schema, 'a, T: AsRef<OptionLinksMetadata>>(
+    supergraph_schema: &FederationSchemaRef<'schema, T>,
+    join_spec_definition: &'a JoinSpecDefinition,
+) -> Result<
+    (
+        FederationSubgraphs,
+        IndexMap<&'schema Name, &'static FederationSpecDefinition>,
+        IndexMap<&'schema Name, &'schema NodeStr>,
```

Comment:
I think this can be gathered in a type, and returned as part of this result. It _theoretically_ should help you simplify these lifetimes too.